File: python/leetcode/2044.py
# https://leetcode.com/problems/count-number-of-maximum-bitwise-or-subsets/


import functools


class Solution:
    # https://leetcode.com/problems/count-number-of-maximum-bitwise-or-subsets/editorial/?envType=daily-question&envId=2025-07-28#approach-2-memoization

    def __count_subsets(
        self,
        memo: dict[int, dict[int, int]],
        nums: list[int],
        index: int,
        current_or: int,
        target_or: int,
    ) -> int:
        if index in memo and current_or in memo[index]:
            return memo[index][current_or]

        if index == len(nums):
            return 1 if current_or == target_or else 0

        if not index in memo:
            memo[index] = {}

        memo[index][current_or] = self.__count_subsets(
            memo, nums, index + 1, current_or, target_or
        ) + self.__count_subsets(
            memo, nums, index + 1, current_or | nums[index], target_or
        )

        return memo[index][current_or]

    def countMaxOrSubsets(self, nums: list[int]) -> int:
        memo: dict[int, dict[int, int]] = {}
        target_or = functools.reduce(lambda x, y: x | y, nums)

        return self.__count_subsets(memo, nums, 0, 0, target_or)


# TODO: check after dp

if __name__ == "__main__":
    test = [2, 2, 2]

    print(Solution().countMaxOrSubsets(test))

File: golang/leetcode/2044.go
package leetcode

func countSubsets(
	memo map[int]map[int]int,
	nums []int,
	index int,
	current_or int,
	target_or int,
) int {
	if index == len(nums) {
		if current_or == target_or {
			return 1
		}

		return 0
	}

	if _, exists := memo[index]; !exists {
		memo[index] = make(map[int]int)
	} else if val, exists := memo[index][current_or]; exists {
		return val
	}

	result := countSubsets(memo, nums, index+1, current_or, target_or) + countSubsets(memo, nums, index+1, current_or|nums[index], target_or)

	memo[index][current_or] = result

	return result
}

func CountMaxOrSubsets(nums []int) int {
	memo := make(map[int]map[int]int)
	max_v := 0

	for _, v := range nums {
		max_v |= v
	}

	return countSubsets(memo, nums, 0, 0, max_v)
}
